File: integration-tests/testdata/programs/hcloud-js/hetzner.ts
import * as hcloud from "@pulumi/hcloud";
import * as pulumi from "@pulumi/pulumi";
import * as forge from "node-forge";
import { Cluster, DeployedServer } from "./types";

const defaultTalosInitialVersion = "v1.10.3";
const variant = "metal";

export function Hetzner(cluster: Cluster): DeployedServer[] {
  const sshKey = new hcloud.SshKey(
    "ssh",
    {
      publicKey: generateSSHKey().then((keys) => keys.publicKey),
    },
    { ignoreChanges: ["publicKey"] },
  );

  // Create the private network
  const network = new hcloud.Network("private-network", {
    name: pulumi.interpolate`private-network-${cluster.name}`,
    ipRange: cluster.privateNetwork,
  });

  // Convert network ID to integer
  const convertedNetID = network.id.apply((id) => parseInt(id, 10));

  // Add a subnet to the private network
  new hcloud.NetworkSubnet("private-subnet", {
    networkId: convertedNetID,
    type: "server",
    networkZone: "eu-central", // Adjust based on your preferred region
    ipRange: cluster.privateSubnetwork,
  });

  const deployed: DeployedServer[] = [];

  for (const machine of cluster.machines) {
    if (!machine.hcloud) {
      throw new Error(`machine ${machine.id} is missing hcloud configuration`);
    }

    const serverArch = architectureForServer(machine.hcloud.serverType);
    const machineVariant = machine.platform ?? variant;
    const talosVersion = machine.talosInitialVersion ?? defaultTalosInitialVersion;
    const selector = `os=talos,version=${talosVersion},variant=${machineVariant},arch=${serverArch}`;
    const image = hcloud.getImage({
      withSelector: selector,
      withArchitecture: serverArch,
    });

    if (!machine.privateIP) {
      throw new Error(`machine ${machine.id} is missing privateIP`);
    }

    const serverArgs: hcloud.ServerArgs = {
      name: machine.id,
      serverType: machine.hcloud.serverType,
      image: image.then((v) => `${v.id}`),
      sshKeys: [sshKey.id],
      networks: [
        {
          networkId: convertedNetID,
          ip: machine.privateIP,
        },
      ],
    };

    if (machine.hcloud.datacenter) {
      serverArgs.datacenter = machine.hcloud.datacenter;
    } else {
      serverArgs.location = "nbg1";
    }

    // Define the server
    const server = new hcloud.Server(
      machine.id,
      serverArgs,
      { ignoreChanges: ["sshKeys"] },
    );

    deployed.push({
      id: machine.id,
      ip: server.ipv4Address,
    });
  }

  return deployed;
}

function architectureForServer(serverType: string): string {
  if (serverType.startsWith("cax") || serverType.startsWith("cpx")) {
    return "arm";
  }

  return "x86";
}

async function generateSSHKey(): Promise<{ publicKey: string }> {
  return new Promise((resolve, reject) => {
    try {
      // Generate an RSA key pair
      const keypair = forge.pki.rsa.generateKeyPair(2048);

      // Convert to PEM format
      const publicKeyForge = forge.ssh.publicKeyToOpenSSH(keypair.publicKey);

      const publicKey = `${publicKeyForge}`;

      resolve({ publicKey });
    } catch (error) {
      reject(`Error generating SSH keys: ${error}`);
    }
  });
}

File: integration-tests/testdata/programs/hcloud-ha-py/hetzner.py
from __future__ import annotations

from typing import Any, Dict, List

import pulumi
from pulumi_hcloud import (
    Network,
    NetworkSubnet,
    Server,
    SshKey,
    get_image_output,
)

# Static demo key; replace with your own for real runs.
_DEMO_PUBLIC_KEY = (
    "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJtX7iBJ8zZCNdSP6NqBqXex12MNl81pHR38t0KBfZ1f demo@example"
)

def architecture_for_server(server_type: str) -> str:
    if server_type.startswith("cax") or server_type.startswith("cpx"):
        return "arm"
    return "x86"


def hetzner(cluster: Any) -> List[Dict[str, pulumi.Output]]:
    ssh_key = SshKey(
        "ssh",
        public_key=_DEMO_PUBLIC_KEY,
        opts=pulumi.ResourceOptions(ignore_changes=["publicKey"]),
    )

    network = Network(
        "private-network",
        name=pulumi.Output.concat("private-network-", cluster.name),
        ip_range=cluster.privateNetwork,
    )
    converted_net_id = network.id.apply(lambda net_id: int(net_id))

    NetworkSubnet(
        "private-subnet",
        network_id=converted_net_id,
        type="server",
        network_zone="eu-central",
        ip_range=cluster.privateSubnetwork,
    )

    deployed: List[Dict[str, pulumi.Output]] = []

    for machine in cluster.machines:
        if machine.hcloud is None:
            raise ValueError(f"machine {machine.id} is missing hcloud configuration")
        if not machine.privateIP:
            raise ValueError(f"machine {machine.id} is missing privateIP")

        server_arch = architecture_for_server(machine.hcloud.serverType)
        machine_variant = machine.variant
        talos_version = machine.talosInitialVersion
        selector = f"os=talos,version={talos_version},variant={machine_variant},arch={server_arch}"

        image = get_image_output(with_selector=selector, with_architecture=server_arch)

        server_args: Dict[str, Any] = {
            "name": machine.id,
            "server_type": machine.hcloud.serverType,
            "image": image.id.apply(lambda v: str(v)),
            "ssh_keys": [ssh_key.id],
            "networks": [
                {
                    "network_id": converted_net_id,
                    "ip": machine.privateIP,
                }
            ],
        }

        if machine.hcloud.datacenter:
            server_args["datacenter"] = machine.hcloud.datacenter
        else:
            server_args["location"] = "nbg1"

        server = Server(
            resource_name=machine.id,
            opts=pulumi.ResourceOptions(ignore_changes=["sshKeys", "userData"]),
            **server_args,
        )

        deployed.append({
            "id": machine.id,
            "ip": server.ipv4_address,
        })

    return deployed
